components/style/gecko/conversions.rs

```diff
@@ -338,6 +340,182 @@ impl nsStyleImage {
+            },
+            nsStyleImageType::eStyleImageType_Gradient => {
+                Some(GenericImage::Gradient(self.get_gradient()))
+            },
+            nsStyleImageType::eStyleImageType_Element => {
+                use gecko_string_cache::Atom;
+                let atom = Gecko_GetImageElement(self);
+                Some(GenericImage::Element(Atom::from_addrefed(atom)))
+            },
+            x => panic!("Unexpected image type {:?}", x)
+        }
+    }
+
+    unsafe fn get_image_url(self: &nsStyleImage) -> SpecifiedUrl {
+        use gecko_bindings::bindings::Gecko_GetURLValue;
+        let url_value = Gecko_GetURLValue(self);
+        let mut url =
+            SpecifiedUrl::from_url_value_data(url_value.as_ref().unwrap()).expect("self has valid URL value");
+        url.build_image_value();
+        url
+    }
+
+    unsafe fn get_gradient(self: &nsStyleImage) -> Gradient {
+        use gecko::values::convert_nscolor_to_rgba;
+        use gecko_bindings::bindings::Gecko_GetGradientImageValue;
+        use gecko_bindings::structs::{NS_STYLE_GRADIENT_SHAPE_CIRCULAR, NS_STYLE_GRADIENT_SHAPE_ELLIPTICAL};
+        use gecko_bindings::structs::{NS_STYLE_GRADIENT_SHAPE_LINEAR, NS_STYLE_GRADIENT_SIZE_CLOSEST_CORNER};
+        use gecko_bindings::structs::{NS_STYLE_GRADIENT_SIZE_CLOSEST_SIDE, NS_STYLE_GRADIENT_SIZE_EXPLICIT_SIZE};
+        use gecko_bindings::structs::{NS_STYLE_GRADIENT_SIZE_FARTHEST_CORNER, NS_STYLE_GRADIENT_SIZE_FARTHEST_SIDE};
+        use gecko_bindings::structs::root::nsStyleUnit::{eStyleUnit_None, eStyleUnit_Percent};
+        use values::computed::{Length, LengthOrPercentage};
+        use values::computed::image::LineDirection;
+        use values::computed::position::Position;
+        use values::generics::image::{ColorStop, CompatMode, Circle, Ellipse, EndingShape, GradientKind, ShapeExtent};
+        use values::specified::length::Percentage;
+        use values::specified::position::{X, Y};
+
+        let gecko_gradient = Gecko_GetGradientImageValue(self).as_ref().unwrap();
+        let gradient_kind = match gecko_gradient.mShape as u32 {
+            NS_STYLE_GRADIENT_SHAPE_LINEAR => {
+                let line_direction = match *gecko_gradient.mAngle.get_mUnit() {
+                    eStyleUnit_None => {
+                        debug_assert!(*gecko_gradient.mBgPosX.get_mUnit() == eStyleUnit_Percent &&
+                                      *gecko_gradient.mBgPosY.get_mUnit() == eStyleUnit_Percent);
+                        let pos_x = LengthOrPercentage::from_gecko_style_coord(&gecko_gradient.mBgPosX)
+                            .expect("mBgPosX has valid data");
+                        let pos_y = LengthOrPercentage::from_gecko_style_coord(&gecko_gradient.mBgPosY)
+                            .expect("mBgPosY has valid data");
+                        let horizontal_direction = match pos_x {
+                            LengthOrPercentage::Percentage(Percentage(0.0)) => X::Left,
```

Comment:
Matching against float literals is a warning in nightly, please don't pattern match them if you have time.